urllib3/sync_connection.py

```diff
@@ -0,0 +1,556 @@
+# -*- coding: utf-8 -*-
+"""
+This module implements the synchronous connection management logic.
+
+Unlike in http.client, the connection here is an object that is responsible
+for a very small number of tasks:
+
+    1. Serializing/deserializing data to/from the network.
+    2. Being able to do basic parsing of HTTP and maintaining the framing.
+    3. Understanding connection state.
+
+This object knows very little about the semantics of HTTP in terms of how to
+construct HTTP requests and responses. It mostly manages the socket itself.
+"""
+from __future__ import absolute_import
+
+import collections
+import datetime
+import io
+import itertools
+import socket
+import warnings
+
+import h11
+
+from .base import Response
+from .exceptions import (
+    ConnectTimeoutError, NewConnectionError, SubjectAltNameWarning,
+    SystemTimeWarning, BadVersionError, FailedTunnelError
+)
+from .packages import six
+from .util import selectors, connection, ssl_ as ssl_util
+
+try:
+    import ssl
+except ImportError:
+    ssl = None
+
+
+# When updating RECENT_DATE, move it to
+# within two years of the current date, and no
+# earlier than 6 months ago.
+RECENT_DATE = datetime.date(2016, 1, 1)
+
+_SUPPORTED_VERSIONS = frozenset([b'1.0', b'1.1'])
+
+
+def _headers_to_native_string(headers):
+    """
+    A temporary shim to convert received headers to native strings, to match
+    the behaviour of httplib. We will reconsider this later in the process.
+    """
+    # TODO: revisit.
+    # This works because fundamentally we know that all headers coming from
+    # h11 are bytes, so if they aren't of type `str` then we must be on Python
+    # 3 and need to decode the headers using Latin1.
+    for n, v in headers:
+        if not isinstance(n, str):
+            n = n.decode('latin1')
+        if not isinstance(v, str):
+            v = v.decode('latin1')
+        yield (n, v)
+
+
+def _stringify_headers(headers):
+    """
+    A generator that transforms headers so they're suitable for sending by h11.
+    """
+    # TODO: revisit
+    for name, value in headers:
+        if isinstance(name, six.text_type):
+            name = name.encode('ascii')
+
+        if isinstance(value, six.text_type):
+            value = value.encode('latin-1')
+        elif isinstance(value, int):
+            value = str(value).encode('ascii')
```

Comment:
Not on all supported versions of Python 3 iirc. I'm fairly certain that was only added back in 3.5